libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,453 @@
+// recorded without post-editing still tend to achieve lower scores.
+double GetQuantizationDistance(
+   const std::vector<int>& peakIndices, const std::vector<float>& peakValues,
+   size_t size, int numDivisions, int lag)
+{
+   std::vector<double> peakDistances(peakIndices.size());
+   const auto peakSum =
+      std::accumulate(peakValues.begin(), peakValues.end(), 0.);
+   const auto odfSamplesPerDiv = 1. * size / numDivisions;
+   std::transform(
+      peakIndices.begin(), peakIndices.end(), peakDistances.begin(),
+      [&](int peakIndex) {
+         const auto shiftedIndex = peakIndex - lag;
+         const auto closestDiv = std::round(shiftedIndex / odfSamplesPerDiv);
+         // Normalized distance between 0 and 1:
+         const auto distance =
+            (shiftedIndex - closestDiv * odfSamplesPerDiv) / odfSamplesPerDiv;
+         // Mutliply by two such that the error spans `[0, 1)`.
+         return 2 * std::abs(distance);
+      });
+   // Calculate the score as the sum of the distances weighted by
+   // the odf values:
+   const auto weightedAverage =
+      std::inner_product(
+         peakDistances.begin(), peakDistances.end(), peakValues.begin(), 0.) /
+      peakSum;
+   return weightedAverage;
+}
+
+OnsetQuantization RunQuantizationExperiment(
+   const std::vector<float>& odf, const std::vector<int>& peakIndices,
+   const std::vector<float>& peakValues,
+   const std::vector<int>& possibleNumTatums,
+   QuantizationFitDebugOutput* debugOutput)
+{
+   const auto quantizations = [&]() {
+      std::unordered_map<int, OnsetQuantization> quantizations;
+      std::transform(
+         possibleNumTatums.begin(), possibleNumTatums.end(),
+         std::inserter(quantizations, quantizations.end()), [&](int numTatums) {
+            const auto lag = GetOnsetLag(odf, numTatums);
+            const auto distance = GetQuantizationDistance(
+               peakIndices, peakValues, odf.size(), numTatums, lag);
+            return std::make_pair(
+               numTatums, OnsetQuantization { distance, lag, numTatums });
+         });
+      return quantizations;
+   }();
+
+   const auto bestFitIt = std::min_element(
+      quantizations.begin(), quantizations.end(),
+      [](const std::pair<int, OnsetQuantization>& a,
+         const std::pair<int, OnsetQuantization>& b) {
+         return a.second.error < b.second.error;
+      });
+
+   const auto error = bestFitIt->second.error;
+   const auto mostLikelyNumTatums = bestFitIt->first;
+   const auto lag = bestFitIt->second.lag;
+
+   return { error, lag, mostLikelyNumTatums };
+}
+
+std::optional<TimeSignature>
+GetTimeSignature(const BarDivision& barDivision, int numTatums)
+{
+   const auto numBeats = barDivision.numBars * barDivision.beatsPerBar;
+   const auto tatumsPerBeat = 1. * numTatums / numBeats;
+   switch (barDivision.beatsPerBar)
+   {
+   case 2:
+      return (tatumsPerBeat == 3) ? TimeSignature::SixEight :
```

Comment:
I feel a similar doubt here, as with `IsRound` -- you should express the test with integer arithmetic only